### ISSUE TITLE
Epic - improving Ember Data sections

### ISSUE THREAD

**Issue by jenweber** [ember-learn/guides-source#20]
This issue is a reference for future Ember Data section improvements. Following discussion and consensus, these items should be broken into separate issues and small PRs.

- [ ] sample req and res objects/spec for different adapters/serializers
- [ ] Replace "customizing" language for serializers. I think we could simplify the mental model for everyone by teaching more about what a serializer is and how to write one instead of focusing on “customizing”
- [ ] Focus serializer explanations around most useful API methods. There are only 3 necessary methods, only 2 of which are “happy path”. Those methods are `normalizeResponse` `serialize` and `pushPayload`, and each can be described in an ideal world as “a pure function that converts from non-json-api to json-api or vice-versa”. The 2 happy path methods are `normalizeResponse` and `serialize`
- [ ] make sure distinction between `model` and `record` is clear. `record` for instances and `ModelClass` for the class
- [ ] better intro explaining what adapters are
- [ ] overall, gradually align on json-api terminology within ember-data, e.g. instead of saying `record` say `resource` and `resource-identifier`

Helpful terminology definitions:
- `store`: a cache for `records`
- `Model`: a schema class used when instantiating a `record`
- `Adapter`: a request-manager: it takes instructions on what to find and manages the process of making an API request and returning the response. Essentially an abstraction over fetch.
- `Serializer`: a formatter used to convert your API responses into json-api format if they are not already (caveat, needs to be a subset of json-api (camelCase member names and dasherized, singularized types)
- `finders`: a utility method for chaining together `request => format data => update cache`
- `records` (e.g. instances of Models) - think of them as “remote state”. A `record` is some data that lives outside of the lifecycle of the application, e.g. on your server

**Comment by jenweber** [comment]
cc @runspired - please edit however you want ^^^

**Comment by lifeart** [comment]
In serializers: `_normalizeDocumentHelper`, `extractAttributes`, `extractRelationships` - most used methods for me, it will be nice to cover it for newbies

**Comment by ppcano** [comment]
Related [discuss thread](https://discuss.emberjs.com/t/when-to-add-ember-data/15050)

**Comment by zachgarwood** [comment]
I think one incremental change that could be made is to introduce using `fetch` (or some third-party library) to retrieve data in the Super Rentals tutorial. Currently, in part 5 of the tutorial, [Installing Addons](https://guides.emberjs.com/v3.5.0/tutorial/installing-addons/), we have the reader install Mirage, add the dummy data to the mirage config, then tell reader that they won't actually use this data until part 8, [Using Ember Data](https://guides.emberjs.com/v3.5.0/tutorial/ember-data/). In the meantime, we have the reader using hard-coded data that we had them add when we were explaining the `model()` hook in part 4, [The Model Hook](https://guides.emberjs.com/v3.5.0/tutorial/model-hook/).

I propose that immediately after we have the reader install Mirage (in part 5), we have them replace the hard-coded dummy data with a `fetch` call to retrieve data from Mirage in part 6, [Building a Simple Component](https://guides.emberjs.com/v3.5.0/tutorial/simple-component/). I believe this would help demonstrate the path to eventually incorporating Ember Data as an app grows.

@jenweber, if you think this would be an appropriate change, I'd be happy to submit a PR.

**Comment by sukima** [comment]
If I (or others) were to work on this would it be best to use [ember-ajax](https://github.com/ember-cli/ember-ajax) of [`fetch()`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)?

I'm inclined to say ember-ajax as it is included in ember apps by default, has better support for errors, and has good patterns for customization (which is a common need in apps). On the flip side `fetch` is a known JS standard and many people coming to a framework to evaluate its worthy-ness tend to dislike more abstractions over a standard-lib even if said lib lacks features.

Thoughts?

**Comment by zachgarwood** [comment]
@sukima I like the suggestion of using ember-ajax, but I think going from fetch -> ember-ajax -> ember-data may be _too_ gradual of a progression. I think using ember-ajax while stating something like "You may choose to use fetch(), but ember-ajax is included in a standard project and gives us additional functionality, so we're going to stick with that." We might even be able to point out a place or two where we use a feature of ember-ajax that does not appear in fetch, to help sell it.

**Comment by jenweber** [comment]
Sorry for the delay, but I think this should be a topic at this week’s learning team meeting to confirm, before you do any work. @zachgarwood

**Comment by toddjordan** [comment]
Hi! Love seeing energy around the tutorial! There's an issue around use of mirage here where we replace mirage with a publicly hosted tutorial api #229 . (mirage would still be utilized for testing)

IF we were to introduce using standard http apis before ember data in the tutorial, it would probably need to go as follows:

1. we should use ember-ajax instead of fetch.  Fetch doesn't have ie 11 support natively (among other browser related quirks), and we don't want to introduce another addon to the mix. (ember-ajax comes out of the box)
1. network calls are not introduced until the ember data section, so we would need to replace that section with the ajax version of it and possibly add a new section to the tutorial possibly near the end right before deploying.
1. mirage is only mentioned in the addons section because it is an addon we will use later.  It doesn't necessarily have to be there and can be pushed back to the the first use of a network connection. (its install/setup could be pushed into the test part of the section near the end, allowing users to skip if they please, once we move towards using the public api)

I'd also like to mention we are planning a coordinated effort in December to enhance several parts of the Learning infrastructure and for sure tutorial will be a part.  If folks want to get started earlier we have a few guides-source issues here marked with the tutorial label.  Get in touch with me or comment in a relevant issue if you are interested.

https://github.com/ember-learn/guides-source/issues?q=is%3Aissue+is%3Aopen+label%3Atutorial

Like Jen said we'll also be discussing Thursday, so folks are welcome to join then too.

**Comment by zachgarwood** [comment]
@jenweber & @toddjordan any updates?

**Comment by runspired** [comment]
you may want to watch this RFC https://github.com/emberjs/rfcs/pull/395

**Comment by runspired** [comment]
And you may want to consider the effects of `defaultStore` on the learning story, and the need for us to deprecate it to clean things up: https://github.com/emberjs/rfcs/issues/377

**Comment by jenweber** [comment]
I opened a PR for #708 that introduces the use of ember-ajax. Once that goes through, we should examine all the other places in the Guides that Ember Data is used, and see if they could benefit from some ember-ajax examples.

We also plan to do another tutorial, in addition to Super Rentals, that does not use Ember Data. In a way, then, Super Rentals can kinda serve as a better "Ember Data" tutorial, since people will not be learning Ember basics and Ember Data basics at the same time, if they follow the tutorials in order.

This new tutorial project was started a long time ago but never finished. I'm going to bring it up at some learning team meetings soon.
